compressors/lz77.py

```diff
@@ -0,0 +1,705 @@
+        with Uint8FileDataStream(input_file_path, "rb") as fds:
+            with EncodedBlockWriter(encoded_file_path) as writer:
+                self.encode(fds, block_size=block_size, encode_writer=writer)
+
+
+class LZ77Decoder(DataDecoder):
+    def __init__(self, window_initialization: List = None):
+        """Initialize LZ77 decoder.
+
+        Args:
+            window_initialization (List, optional): initialize window (this is like side information
+                or dictionary in zstd parlance). The same initialization should be used as in encoder.
+        """
+        self.window = []
+        if window_initialization is not None:
+            self.window = list(window_initialization)
+
+    def decode_lz77_sequences(self, encoded_bitarray: BitArray):
+        """Perform entropy decoding of the LZ77 sequences and return the decoded sequences
+        and the number of bits consumed.
+
+        Args:
+            encoded_bitarray (BitArray): encoded bit array
+        """
+        # first get the encoded block size so we can pick out the block for decoding
+        encoded_block_size_bitarray = encoded_bitarray[:ENCODED_BLOCK_SIZE_HEADER_BITS]
+        encoded_block_size = bitarray_to_uint(encoded_block_size_bitarray)
+        encoded_bitarray = encoded_bitarray[
+            ENCODED_BLOCK_SIZE_HEADER_BITS : ENCODED_BLOCK_SIZE_HEADER_BITS + encoded_block_size
+        ]
+        # decoded the combined list with literal counts, match lengths and offsets
+        combined_decoded, num_bits_consumed_encoding = EliasDeltaUintDecoder().decode_block(
+            encoded_bitarray
+        )
+        combined_decoded = combined_decoded.data_list
+        assert num_bits_consumed_encoding == encoded_block_size
+        num_bits_consumed = ENCODED_BLOCK_SIZE_HEADER_BITS + encoded_block_size
+        assert (len(combined_decoded) - 1) % 3 == 0
+        num_sequences = (len(combined_decoded) - 1) // 3
+        min_match_length = combined_decoded[0]
+        literal_counts = combined_decoded[1 : 1 + num_sequences]
+        # need to adjust lengths and offsets to undo the processing we did on the compressor
+        match_lengths = [
+            l + min_match_length
+            for l in combined_decoded[1 + num_sequences : 1 + 2 * num_sequences]
+        ]
+        match_offsets = [
+            l + 1 for l in combined_decoded[1 + 2 * num_sequences : 1 + 3 * num_sequences]
+        ]
+        lz77_sequences = [
+            LZ77Sequence(l[0], l[1], l[2])
+            for l in zip(literal_counts, match_lengths, match_offsets)
+        ]
+        return lz77_sequences, num_bits_consumed
+
+    def decode_literals(self, encoded_bitarray: BitArray):
+        """Perform entropy decoding of the literals and return the literals
+        and the number of bits consumed.
+
+        Args:
+            encoded_bitarray (BitArray): encoded bit array
+        """
+        num_bits_consumed = 0
+        # first read the size of the counts encoding
```

Comment:
maybe call this symbol counts (or frequencies), so that it is clear what counts these are..

Reply:
done now